Integrate Core App into controller

grondinjc opened the pull request:
The solution is not fully integrated since it needs to be integrated into frontend

Commit:
Adding core files (WIP)

Example of execution :
python core.py dummyProjectName /tmp/test/dummyProjectName

Commit:
Added basic app without threads

ZoneTransit::add had to be changed a bit to make it work ...
When call the function with an element at the highest postion
it does not get inserted in the vector ...

Commit:
Added threadpool for async tasks

Commit:
Added doc message

Commit:
Removed old file before mv

Commit:
Fixed buggy change

Commit:
Added EOF line break

Commit:
WIP - Starting to integrate app in controller

Also created file path verification

Commit:
Added parameter validation and call to app

Commit:
Adding /ide/tree to fetch project nodes

Commit:
Remove debug code - Still need to update frontend

grondinjc commented:
Async callback are not added yet. I'm working on this.

grondinjc commented:
@vphilippon Check this for feedbacks please
(no need to merge, yet)

maym2104 commented on src/app/python/core.py:
```diff
@@ -0,0 +1,230 @@
+                            Suppression as EditRemove)
+
+# Class to hold every change element using a name
+# Data field would be the conent if is_add is True and the count when False
```
The count being the size of the deletion?

grondinjc replied:
Yes

Commit:
Added notification stategy to avoid possible loop

vphilippon commented on src/app/python/core.py:
```diff
@@ -0,0 +1,230 @@
+    @type num_threads: int
+
+    @param project_name: The CIDE.py logger instance
+    @param project_path: The CIDE.py logger instance
```
Change de params description.
Little copy-paste error.

vphilippon commented on src/server/ideController.py:
```diff
@@ -5,25 +6,54 @@
     """
     IDEController initialiser
 
+    @param app: The core application
```
Add the `@type cide.app.python.core` type in docstring if possible. That's nice for people using autocompletion (like me!)

vphilippon commented:
@grondinjc Pretty nice, added a few docsting notes.

Commit:
Added docstrings

Commit:
Applied function translation

Commit:
Added callback to controller to get notifications

grondinjc commented on bin/startCIDE.py:
```diff
@@ -46,10 +47,11 @@
 
 # Instanciate App
 chatApp = Chat(logger)
+coreApp = Core("dummyProjectName", "/tmp/test/dummyProjectName", logger, 4)
```
@vphilippon should I use a different logger instance ?

vphilippon replied:
@grondinjc Not really, since we'll probably use the same config and file.

vphilippon replied:
@grondinjc The project name should be set in a config file (May be done in next version)

Commit:
Added unit tests, full loop almost complete

Commit:
Fixed path validation

Python2.7 strings can be either unicode type or str type
Both must be checked when validating a parameter

Commit:
Full loop complete - Missing applied changes from c++

Commit:
Remove blank lines

grondinjc commented:
@vphilippon ready for review

vphilippon commented on src/server/ideController.py:
```diff
@@ -6,26 +7,101 @@
+            (not path.endswith('/')))
+
+  @staticmethod
+  def is_valid_changes(changes):
```
Is it really needed? Let's think about it

vphilippon commented on src/server/ideController.py:
```diff
@@ -6,26 +7,101 @@
 
+# Will be changed to httpErrors
+def create_argument_error_msg(arg):
+  return {'code': 400, 
```
Use `cherrypy.response.status` or `raise HttpError` instead

vphilippon commented:
There are a few things we'll want to change latter, but for the sake of going forward: :+1:
